### PR TITLE
Virtual tutor merge lecture and exercise

### PR DESCRIPTION
**Pull request by magnemo**
## Description

This pull request contains these completed PBI's:

[Web Create] As a content creator, I want to be able to put an exercise on the section I'm creating
[Web Create] As a content creator, I want to be able to put a lecture on the section I'm creating
[Web Edit] As a content creator, I want to be able to edit or delete sections, lectures or exercises of the courses I haven't published yet

## Changes

All changes have been made in the section edit page (excluding bug fixes).

General changes
Translated to Portuguese.
A limit of having max 10 exercises + lectures has been implemented.

- Exercises
  - You can now create exercises.
  - You can now see a list of existing exercises.
  - You can now update details of an existing exercise.
  - You can now delete exercises.
  - Exercise answers can be added or removed.
  - Exercise can now have 2-4 answers, where only 1 can be correct.

- Lectures
  - You can now create lectures.
  - You can now see a list of existing lectures.
  - You can now update details of an existing lecture.
  - You can now delete lectures.
  - Lecture creating has a visual drop-zone for file upload. (functionality not implemented yet)
  - A limit of having max 7 lectures has been implemented.

Bug fixes:
Courses (Create/Update) now sends a number value for difficulty.
Course (Create/Update) now sends a translated response of the categories to the back end.

## Checklist

- [x] Code has been tested locally and passes all relevant tests.
- [x] Documentation has been updated to reflect the changes, if applicable.
- [x] Code follows the established coding style and guidelines of the project.
- [x] All new and existing tests related to the changes have passed.
- [x] Any necessary dependencies or new packages have been properly documented.
- [x] Pull request title and description are clear and descriptive.
- [x] Reviewers have been assigned to the pull request.
- [x] Any potential security implications have been considered and addressed.
- [x] Performance impact of the changes has been evaluated, if relevant.

## Notes for Reviewers
The backend used for this pull requist is "VirtualTurtor-Merge".
This code needs access to mongoDB to function.

Linting gives a lot of errors, we fixed a lot of the existing errors linting gave, but the rest are non trivial errors that don't inhibit the code and are therefore not a priority right no

### CODE REVIEW

**Review by PrometheusBork — APPROVED** [review]
Looks fine to me. Good work!

**Review by PrometheusBork — CHANGES_REQUESTED** [review]
Code itself looks fine but it does not pass github check

**Review by MortenMunk — CHANGES_REQUESTED** [review]
This does not pass linting:
```
src/services/auth.services.ts(31,1): error TS1128: Declaration or statement expected.
```

**Comment by SlovenlyTF** [comment]
A test on dev is not working. It gets the error: "SyntaxError: Cannot use 'import.meta' outside a module"...
Not our test, and **the code is working just fine outside the test file**.
It's just because the test file isn't considered by Node to be a module.
As it isn't our test, we don't know how it is supposed to run and how then to modify it to work on short notice.

**Comment by MortenMunk** [comment]
> A test on dev is not working. It gets the error: "SyntaxError: Cannot use 'import.meta' outside a module"... Not our test, and **the code is working just fine outside the test file**. It's just because the test file isn't considered by Node to be a module. As it isn't our test, we don't know how it is supposed to run and how then to modify it to work on short notice.

The tests should have passed when they were pushed to dev originally, hence the tests not passing now does not make sense, unless conflict has been introduced. I think we should attempt to fix the tests rather than just ignore them, as this defeats the purpose of tests.... Otherwise we should discuss this with the integration team before merging imo.

File: app/src/components/Exercise/CreateExercisePopUp.tsx
Comment (Adachigaming) [inline app/src/components/Exercise/CreateExercisePopUp.tsx:48]: For now, to get the token you can write `const token = localStorage.getItem("token")`
Don't know how important it is for your PBIs, but still.

Comment (magnemo) [reply app/src/components/Exercise/CreateExercisePopUp.tsx:48]: Hey have checked it out, our code is not set up to take a null value, which the "localStorage.getItem("token")" can return so for now we will not use it.

**Review by PrometheusBork — APPROVED** [review]
Looks good!

File: app/src/pages/SectionEdit.tsx
Comment (JackSMunn) [inline app/src/pages/SectionEdit.tsx:74]: Endpoint does not meet API guidelines.

File: app/src/pages/SectionEdit.tsx
Comment (JackSMunn) [inline app/src/pages/SectionEdit.tsx]: Endpoint does not meet API guidelines.

File: app/src/services/auth.services.ts
Comment (JackSMunn) [inline app/src/services/auth.services.ts:58]: Is this reason sufficient? Is it used to show something to developers or users?

File: app/src/services/auth.services.ts
Comment (JackSMunn) [inline app/src/services/auth.services.ts:65]: Is this reason sufficient? Is it used to show something to developers or users?

File: app/src/services/course.services.ts
Comment (JackSMunn) [inline app/src/services/course.services.ts:63]: Endpoint does not meet API guidelines

**Review by JackSMunn — CHANGES_REQUESTED** [review]
GJ on code!

The requested changes are mostly to address:

* Missing token validation
* Some endpoints do not meet API guidelines

File: app/src/services/auth.services.ts
Comment (SlovenlyTF) [reply app/src/services/auth.services.ts:58]: No idea.
It's onboardings code, we just fixed the hardcoded URL.
The (") insted of (') are an error autofix from a VS code extension.
We'll remove the extension in the future.

File: app/src/services/auth.services.ts
Comment (SlovenlyTF) [reply app/src/services/auth.services.ts:65]: No idea.
It's onboardings code, we just fixed the hardcoded URL.
The (") insted of (') are an error autofix from a VS code extension.
We'll remove the extension in the future.

File: app/src/pages/SectionEdit.tsx
Comment (SlovenlyTF) [reply app/src/pages/SectionEdit.tsx]: Have changed it from api/exercises/getall/${sid} --> api/exercises/section/${sid}
It's the best solution without changing every get request last minute with the query setup.
There's already one without "getall", which is why there is "section"

File: app/src/pages/SectionEdit.tsx
Comment (SlovenlyTF) [reply app/src/pages/SectionEdit.tsx]: Have changed it from api/lectures/getall/${sid} --> api/lectures/section/${sid}
It's the best solution without changing every get request last minute with the query setup.
There's already one without "getall", which is why there is "section"

File: app/src/services/course.services.ts
Comment (SlovenlyTF) [reply app/src/services/course.services.ts]: Have fixed. Is now just "/api/courses"

**Review by JackSMunn — APPROVED** [review]

**Comment by JackSMunn** [comment]
I have approved, however I have not tested the frontend locally, so you will need approvals that have done so before merging.

**Review by PrometheusBork — APPROVED** [review]
Looks good

**Review by MortenMunk — APPROVED** [review]
Looks good for now. We should look at commented code, but I know this is in agreement with Frederik for this PR, so I approve 👍 🥇

**Pull request merged**
